hail/python/test/hailtop/aiotools/test_copy.py

```diff
@@ -0,0 +1,351 @@
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x', treat_dest_as=Transfer.TARGET_FILE))
+
+    await expect_file(fs, f'{dest_base}x', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_overwrite_rename_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+    await create_test_file(fs, 'dest', dest_base, 'x')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x'))
+
+    await expect_file(fs, f'{dest_base}x', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_rename_dir(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_dir(fs, 'src', src_base, 'a/')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x'))
+
+    await expect_file(fs, f'{dest_base}x/file1', 'src/a/file1')
+    await expect_file(fs, f'{dest_base}x/subdir/file2', 'src/a/subdir/file2')
+
+
+@pytest.mark.asyncio
+async def test_copy_file_dest_trailing_slash_target_dir(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base, treat_dest_as=Transfer.TARGET_DIR))
+
+    await expect_file(fs, f'{dest_base}a', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_file_dest_target_dir(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/'), treat_dest_as=Transfer.TARGET_DIR))
+
+    await expect_file(fs, f'{dest_base}a', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_file_dest_target_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}a', treat_dest_as=Transfer.TARGET_FILE))
+
+    await expect_file(fs, f'{dest_base}a', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_overwrite_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+    await create_test_file(fs, 'dest', dest_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/')))
+
+    await expect_file(fs, f'{dest_base}a', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_file_src_trailing_slash(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    with pytest.raises(FileNotFoundError):
+        await fs.copy(Transfer(f'{src_base}a/', dest_base))
+
+
+@pytest.mark.asyncio
+async def test_copy_dir(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_dir(fs, 'src', src_base, 'a/')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/')))
+
+    await expect_file(fs, f'{dest_base}a/file1', 'src/a/file1')
```

Comment:
Is there ever a case where we're expecting to get 

```
{dest_base}file1
{dest_base}subdir
```

Basically putting the contents of a/ directly inside {dest_base}?

Reply:
I added a test_copy_src_parts test for this that copies from multiple sources.  I think this would normally be handled by globbing which would run a layer above the copy function.